ocaml/tests/test_vm_check_operation_error.ml

```diff
@@ -155,6 +107,71 @@ let test_sxm_allowed_when_rum () =
         )
   )
 
+let test_is_allowed_concurrently (expected, (op, current_ops)) =
+  let ops_to_str ops =
+    String.concat "," (List.map (fun (_, op) -> vm_op_to_string op) ops)
+  in
+  let name =
+    match current_ops with
+    | [] ->
+        vm_op_to_string op
+    | lst ->
+        Printf.sprintf "%a when %s" pp_vm_op op (ops_to_str lst)
+  in
+
+  let test () =
+    let actual = Xapi_vm_lifecycle.is_allowed_concurrently ~op ~current_ops in
+    let name =
+      Printf.sprintf "%a allowed in [%s]" pp_vm_op op (ops_to_str current_ops)
+    in
+    Alcotest.(check bool) name expected actual
+  in
+  (name, `Quick, test)
+
+let allowed_specs =
+  let current_of op = ((), op) in
+  let allow_hard_shutdown =
+    List.map
+      (fun op ->
+        let allowed = match op with `hard_shutdown -> false | _ -> true in
+        (allowed, (`hard_shutdown, [current_of op]))
+      )
+      API.vm_operations__all
+  in
+  let allow_hard_reboot =
+    List.map
+      (fun op ->
+        let allowed =
+          match op with `hard_shutdown | `hard_reboot -> false | _ -> true
+        in
+        (allowed, (`hard_reboot, [current_of op]))
+      )
+      API.vm_operations__all
+  in
+  let allow_clean_shutdown =
+    List.map
+      (fun op ->
+        let allowed = match op with `migrate_send -> true | _ -> false in
+        (allowed, (`clean_shutdown, [current_of op]))
+      )
+      API.vm_operations__all
+  in
+  List.concat
+    [
+      [
+        (true, (`snapshot, []))
+      ; (true, (`snapshot, [current_of `checkpoint]))
+      ; (false, (`migrate_send, [current_of `clean_reboot]))
+      ; (true, (`clean_reboot, [current_of `migrate_send]))
+      ]
+    ; allow_hard_shutdown
+    ; allow_clean_shutdown
+    ; allow_hard_reboot
+    ]
```

Comment:
this function to me looks quite similar to the `is_allowed_concurrently` implementation (another attempt of encoding what is allowed and what is not). I think a proper data structure would be good to encode this kind of relationships between different operations, such as a graph. And the problem of whether operations are allowed to coexist would become a search in the graph to see if there is an edge in it.

Reply:
But this might be an overkill

Reply:
A graph might be worth for general-case testing, here I was interested in testing a particular set of cases, while also having some rules to generate them. I didn't want to spend more time on this. If there are issues around the area, testing can be reinforced using graphs or better ways to encode properties

Reply:
I think the idea of a graph would be useful in the actual business logic, to make it easier to encode rules. For the unit tests, we'd directly test for interesting cases without clever logic.